runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/BeamWorkerStatusGrpcService.java

```diff
@@ -0,0 +1,190 @@
+    return new BeamWorkerStatusGrpcService(apiServiceDescriptor, headerAccessor);
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (CompletableFuture<WorkerStatusClient> clientFuture : connectedClient.values()) {
+      if (clientFuture.isDone()) {
+        clientFuture.get().close();
+      }
+    }
+    connectedClient.clear();
+  }
+
+  @Override
+  public StreamObserver<WorkerStatusResponse> workerStatus(
+      StreamObserver<WorkerStatusRequest> requestObserver) {
+    String workerId = headerAccessor.getSdkWorkerId();
+    LOG.info("Beam Fn Status client connected with id {}", workerId);
+
+    WorkerStatusClient fnApiStatusClient =
+        WorkerStatusClient.forRequestObserver(workerId, requestObserver);
+    fnApiStatusClient.setDeregisterCallback(this.connectedClient::remove);
+    if (connectedClient.containsKey(workerId) && connectedClient.get(workerId).isDone()) {
+      LOG.info(
+          "SDK Worker {} was connected to status server previously, disconnecting the old client",
+          workerId);
+      try {
+        WorkerStatusClient oldClient = connectedClient.get(workerId).get();
+        oldClient.close();
+      } catch (IOException | InterruptedException | ExecutionException e) {
+        LOG.warn("Error closing worker status client", e);
+      }
+    }
+    connectedClient
+        .computeIfAbsent(workerId, k -> new CompletableFuture<>())
+        .complete(fnApiStatusClient);
+    return fnApiStatusClient.getResponseObserver();
+  }
+
+  /**
+   * Get the latest SDK worker status from the client's corresponding SDK Harness.
+   *
+   * @param workerId worker id of the SDK harness.
+   * @return {@link CompletableFuture} of WorkerStatusResponse from SDK harness.
+   */
+  public CompletableFuture<WorkerStatusResponse> getWorkerStatus(String workerId) {
+    try {
+      return getStatusClient(workerId).getWorkerStatus();
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      if (e instanceof InterruptedException) {
+        Thread.currentThread().interrupt();
+      }
+      CompletableFuture<WorkerStatusResponse> error = new CompletableFuture<>();
+      error.completeExceptionally(e);
+      return error;
+    }
+  }
+
+  /**
+   * Get all the statuses from all connected SDK harnesses within specified timeout.
+   *
+   * @param timeout max time waiting for the response from all SDK harness.
+   * @param timeUnit timeout time unit.
+   * @return All the statuses in a Map which key is the SDK harness id and value is the status
+   *     content.
+   */
+  public Map<String, String> getAllWorkerStatuses(long timeout, TimeUnit timeUnit) {
+    // return result in worker id sorted map.
+    Map<String, String> allStatuses = new ConcurrentSkipListMap<>(Comparator.naturalOrder());
+
+    List<CompletableFuture<WorkerStatusResponse>> responses = new ArrayList<>();
+    for (String id : connectedClient.keySet()) {
+      responses.add(
+          getWorkerStatus(id)
```

Comment:
The current logic here will block till the SDK harness connects. It might be better if we just populate the map with "Error: Not connected." for any SDK harness that isn't connected.

Reply:
done.